### PR TITLE
Skip AnalyzerSilence if intro/outro cues exist

### PR DESCRIPTION
**Pull request by uklotzde**
This fixes the first and most obvious flaw I found.

The intro/outro cues are re-analyzed even if they have already been set by the analyzer. Both cues must be set manually to skip the analyzer.

Now the track is only processed if at least one of the intro start or outro end cues are settable. The main cue point must not be considered for this decision. It is only set conditionally if the intro start cue is also set.

@ninomp Please verify!

...a bigger refactoring of the Analyzer interface and framework will follow. Not only `AnalyzerSilence` is executed more frequently than it should.

### CODE REVIEW

File: src/analyzer/analyzersilence.cpp [hunk 1]
```diff
@@ -9,18 +9,31 @@ constexpr float kSilenceThreshold = 0.001;
 // TODO: Change the above line to:
 //constexpr float kSilenceThreshold = db2ratio(-60.0f);
 
-}  // anonymous namespace
+bool shouldUpdateMainCue(CuePosition cue) {
```
Comment (daschuer) [inline src/analyzer/analyzersilence.cpp]: The function makes only sense if we pass a main cue, right?

So I think we should rename
cue -> mainCue

File: src/analyzer/analyzersilence.cpp [hunk 2]
```diff
@@ -9,18 +9,31 @@ constexpr float kSilenceThreshold = 0.001;
...
+    return cue.getSource() != Cue::MANUAL || cue.getPosition() == -1.0 || cue.getPosition() == 0.0;
+}
+
+bool hasIntroCueStart(const Cue& cue) {
```
Comment (daschuer) [inline src/analyzer/analyzersilence.cpp]: Is this function unrelated to into and outro?
Than we should omit intro in the function name.

File: src/analyzer/analyzersilence.cpp [hunk 3]
```diff
@@ -9,18 +9,31 @@ constexpr float kSilenceThreshold = 0.001;
...
+    return cue.getPosition() != -1.0;
+}
+
+bool hasOutroCueEnd(const Cue& cue) {
```
Comment (daschuer) [inline src/analyzer/analyzersilence.cpp]: Same here: Can we omit outro in the function name?

**Review by daschuer — CHANGES_REQUESTED** [review]
I have added some naming nits. The rest LGTM. Thank you for taking care.

File: src/analyzer/analyzersilence.cpp (hunk 2 above)
Comment (uklotzde) [reply src/analyzer/analyzersilence.cpp]: These position values are specific for the intro/outro cues. The encoding is not completely unambiguous and unfortunately scattered everywhere in the code.

The way how cue points are currently stored is suboptimal for representing both single points and partial or complete ranges. Those requirements were unknown at the time the data model was designed. This should be considered for a future version of the library.

File: src/analyzer/analyzersilence.cpp (hunk 3 above)
Comment (uklotzde) [reply src/analyzer/analyzersilence.cpp]: Same here. Let's keep it as is.

**Comment by daschuer** [comment]
Thank you.

**Comment by wiezo** [comment]
Thanks :-)
windows version on http://downloads.mixxx.org/builds/appveyor/mixxx-2.3.0-alpha-pre-master-git6823-release-x64.exe

**Comment by wiezo** [comment]
The bug of projection the fade end beyond the track end appeared again. Now 4 times in 21 hours while most tracks already analyzed.

**Comment by daschuer** [comment]
Thank you. 

this unfortunately fails to build on appVeyor's Ubuntu: 

```
src/analyzer/analyzergain.cpp: In member function ‘virtual void AnalyzerGain::storeResults(TrackPointer)’:
src/analyzer/analyzergain.cpp:68:77: error: no match for ‘operator<<’ (operand types are ‘QDebug’ and ‘QFileInfo’)
     qDebug() << "ReplayGain 1.0 result is" << fReplayGainOutput << "dB for" << tio->getFileInfo();
In file included from /usr/include/x86_64-linux-gnu/qt5/QtCore/QtDebug:1:0,
                 from src/analyzer/analyzergain.cpp:2:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qdebug.h:136:20: note: candidate: QDebug& QDebug::operator<<(QChar)
     inline QDebug &operator<<(QChar t) { putUcs4(t.unicode()); return maybeSpace(); }
```

**Comment by uklotzde** [comment]
The answer belongs here.

@wiezo The bug you describe cannot be caused by these recent changes. The intro/outo cue positions didn't change at all and the analyzer will modify them less frequently then before.

Please open a separate bug and describe in detail what does not work as intended, maybe including some screenshot and the exact positions of the cue points that got mixed up.

**Pull request merged**